### PR TITLE
Engine and model-layer refactor

### PR DESCRIPTION
**Pull request by haok1402**
# Engine and model-layer refactor

A cleanup of the engine and model layer. The training recipe is unchanged; the work removes plumbing that only forwarded arguments around and tightens the interfaces between the models, the pipeline, and the runtime.

## Changes

**Runtime contexts.** Process-global state (distributed groups, the model and optimizers, the logger) lives in per-concern context modules, each filled in once at startup by a `setup_*` function and read directly where needed. This replaces the per-task `Ctx` objects that were threaded through every call and the `*_context` managers that wrapped setup.

**Config passed down, not expanded early.** Modules take the transformer config object and read the fields they need locally, instead of callers unpacking scalars and threading them through constructors. Fields are pulled out only where a value genuinely differs from the config (an override or a derived size), which keeps each module's dependencies visible at its own definition.

**Decoder-layer stage naming.** The layer's three compute entry points are renamed to `forward_stage1/3/5`, matching the pipeline's five-stage attention → dispatch → MLP → combine → aggregate split (stages 2 and 4 are the engine-driven all-to-alls). The routing metadata that stage 1 produces and stage 5 consumes now travels as one named record instead of a bare tuple.

**Engine consolidation.** The autograd wrappers and dispatch/combine helpers moved in alongside the stage implementations, so the forward/backward flow reads in one place. The overlap loop and layer partitioning were simplified in the same pass.

**Variable-sequence-length interface.** The pipeline can carry per-document boundaries through to attention and the position embedding, the groundwork for packed / variable-length (SFT) training. The residual stream stays batch-of-one BSHD; only attention sees the flattened view, and with no boundaries set the behavior is identical to before. No entrypoint produces packed batches yet (the corpus-packing data pipeline is a follow-up); the framework side is checked with synthetic boundaries in the multi-GPU test.

**Cleanups.** Each model defines its own stage methods and keeps its helpers local to its file, so it reads top to bottom on its own; DeepSeek dropped its "lite" naming for the base family; run logs unified under one directory; the multi-GPU test renamed; wandb off by default in examples; docs updated.

## Validation

Compared against the pre-refactor baseline on two models spanning different model families, parallelism meshes, and load-balance strategies, both at a constant learning rate. The base-vs-feature load-balance loss curves are on wandb:

| Model | Mesh | Load-balance | wandb (base vs feature) |
|---|---|---|---|
| Qwen3-30B-A3B | PP4 · EP8, 4 nodes | global-batch | [wandb report](https://wandb.ai/pithtrain/pr70?nw=8lddil997mp) |
| DeepSeek-V2-Lite | PP2 · EP2 · DP2, 1 node | sequence | [wandb report](https://wandb.ai/pithtrain/pr70?nw=baq11ybjo6v) |

The curves match: load-balance loss rises, turns over, and decays on both, and cross-entropy agrees within bf16 run-to-run noise.

### CODE REVIEW

File: pithtrain/dualpipe/dualpipev.py [hunk 1]
```diff
@@ -138,6 +161,41 @@ def _reset_states(self) -> None:
...
+            )
+            return
+
+        # Multiple ranks: broadcast the row width first so every stage learns whether this step is
+        # packed (width > 0) or dense (width == 0) before allocating a receive buffer.
+        device = distributed.device
+        width = torch.zeros(1, device=device, dtype=torch.long)
+        if self.is_first_pp_rank and cu_seqlens is not None:
+            width.fill_(cu_seqlens.shape[1])
+        torch.distributed.broadcast(width, group_src=0, group=self.pp_group)
```
[the remark this answers is not in the document]
Comment (haok1402) [reply pithtrain/dualpipe/dualpipev.py:185]: You are wrong. See

https://docs.pytorch.org/docs/2.13/distributed.html#torch.distributed.broadcast

File: pithtrain/dualpipe/dualpipev.py [hunk 2]
```diff
@@ -138,6 +161,41 @@ def _reset_states(self) -> None:
...
+            self.cu_seqlens_chunks = None
+            return
+
+        # Packed: broadcast the (num_chunks, width) offsets and split them per micro-batch.
+        buf = (
+            cu_seqlens.to(device=device, dtype=torch.int32)
+            if self.is_first_pp_rank
+            else torch.empty((num_chunks, int(width)), device=device, dtype=torch.int32)
+        )
+        torch.distributed.broadcast(buf, group_src=0, group=self.pp_group)
```
[the remark this answers is not in the document]
Comment (haok1402) [reply pithtrain/dualpipe/dualpipev.py:196]: You are wrong. See

https://docs.pytorch.org/docs/2.13/distributed.html#torch.distributed.broadcast

File: pithtrain/tasks/pretrain_lm.py [hunk 3]
```diff
@@ -383,19 +373,27 @@ def train_step(cfg: PretrainLMCfg, ctx: PretrainLMCtx) -> None:
...
-                p.grad.mul_(scale)
+    # Token-weighted reduction. The criterion returns per-micro-batch loss SUMS, so gradients
+    # accumulate d(sum loss); dividing by the total non-ignored token count yields the correct
+    # token-mean regardless of how tokens split across micro-batches. The count lives on pipeline
+    # rank 0 (the only rank with labels), so broadcast it to every stage for the gradient scale.
+    num_tokens = torch.ones((), device=device)
+    if distributed.pp_rank == 0:
+        num_tokens.fill_((global_labels != -100).sum().clamp_min(1))
+    if distributed.pp_size > 1:
+        torch.distributed.broadcast(num_tokens, group_src=0, group=distributed.pp_group)
```
[the remark this answers is not in the document]
Comment (haok1402) [reply pithtrain/tasks/pretrain_lm.py:384]: You are wrong. See

https://docs.pytorch.org/docs/2.13/distributed.html#torch.distributed.broadcast

**Comment by haok1402** [comment]
As part of the refactor, we've bumped cutlass-dsl >= 4.6.0 and get rid of the hack that avoids redundant `-libs-base` wheel install. See https://github.com/NVIDIA/cutlass/issues/3259#issuecomment-4862413552 for more details.

**Comment by haok1402** [comment]
@claude review

**Comment by haok1402** [comment]
> 1. **`pithtrain/operators/grouped_linear.py` breaks the documented "tests skip gracefully without `deep_gemm`" gotcha.** The new consolidated file does an unconditional, module-top-level `import deep_gemm` (no try/except), merging what used to be a `deep_gemm`-free BF16-only module (`pithtrain/layers/group_linear.py`) with the FP8 code. `tests/test_grouped_linear_correctness.py` imports `GroupedLinear` from it at module scope with no skip guard (unlike sibling files `test_fp8_quantize_kernels.py` / `test_deepgemm_fp8_linear_correctness.py`, which use a `try/except ImportError` + `pytest.mark.skipif` pattern). Result: collecting this test now hard-fails with `ImportError` in any environment without `deep_gemm` installed, contradicting [AGENTS.md's Testing Gotchas](https://github.com/mlc-ai/pith-train/blob/972e712e894a1e23148ddd646121111261315c2a/AGENTS.md#L154): "Tests skip gracefully when `deep_gemm` is not installed or CUDA is unavailable."

Fixed.

**Comment by haok1402** [comment]
> 2\. **New AGENTS.md sentence cites the class name this same PR deletes.** `pithtrain/models/interface.py` renames `DecoderLayerProtocol` → `LayerProtocol` in this PR, but the newly-added AGENTS.md sentence in the "DualPipeV Pipeline" section still reads "...on `DecoderLayerProtocol` (`pithtrain/models/interface.py`)" — stale the moment it's introduced.
>     https://github.com/mlc-ai/pith-train/blob/972e712e894a1e23148ddd646121111261315c2a/AGENTS.md#L84-L88
> 3\. **AGENTS.md's "Key files" list still describes two files this PR deletes.** `pithtrain/dualpipe/modeling.py` and `pithtrain/dualpipe/layer_partition.py` are both deleted (folded into `execution.py`/`dualpipev.py`), but the "Key files:" bullet list under "DualPipeV Pipeline" is left unchanged and still documents them as if they exist.
>     https://github.com/mlc-ai/pith-train/blob/972e712e894a1e23148ddd646121111261315c2a/AGENTS.md#L90-L95

Fixed.

File: pithtrain/models/gpt_oss.py [hunk 4]
```diff
@@ -1,831 +1,344 @@
...
+        match phase:
+            case 0:
+                stage_count = distributed.pp_size * 2
+                stage_index = distributed.pp_rank
+            case 1:
+                stage_count = distributed.pp_size * 2
+                stage_index = stage_count - 1 - distributed.pp_rank
+            case _:
+                stage_count = 1
+                stage_index = 0
```
Comment (MasterJH5574) [inline pithtrain/models/gpt_oss.py:374]: Suggest **Option B**: name the `-1` (non-pipelined reference) case explicitly and make the default fail fast. As written, `case _` silently maps any unexpected `phase` (e.g. `2`) to a single full-model stage; a named `-1` documents intent and the raising default surfaces bad input instead of hiding it.

```suggestion
        match phase:
            case 0:
                stage_count = distributed.pp_size * 2
                stage_index = distributed.pp_rank
            case 1:
                stage_count = distributed.pp_size * 2
                stage_index = stage_count - 1 - distributed.pp_rank
            case -1:
                # non-pipelined: one stage owns the whole model (reference validation)
                stage_count = 1
                stage_index = 0
            case _:
                raise ValueError(f"phase must be 0, 1, or -1, got {phase}")
```

The identical block appears in `deepseek_v2.py`, `qwen35_moe.py`, and `qwen3_moe.py` — worth applying there too for consistency.

File: pithtrain/dualpipe/execution.py [hunk 5]
```diff
@@ -595,12 +506,349 @@ def create_intermediate_tensors_layer() -> IntermediateTensorsLayer:
...
+    for (_, layer), layer_record in zip(module.layers.items(), chunk_record.layers):
+        hidden_states = layer_forward(layer, hidden_states, rotary_posemb, layer_record, cu_seqlens)
+
+    if module.stage_index == module.stage_count - 1:
+        hidden_states = epilog_f(module, hidden_states, chunk_record.epilog)
+
+    return hidden_states
+
+
+def record_backward(
```
Comment (MasterJH5574) [inline pithtrain/dualpipe/execution.py]: Naming nit on `record_forward` / `record_backward`: the primary job of these functions is to **run** the chunk's forward/backward compute (prolog → layers → epilog, and the reverse). Saving the per-stage tensors into `chunk_record` is a side effect that happens along the way, not the main purpose — so leading with `record_` puts the emphasis on the secondary behavior and reads as if recording were the point.

I think the name is trying to contrast with `reference_forward` (which runs the same compute but relies on ordinary autograd instead of stashing tensors for the manual pipeline backward). That's a real distinction worth keeping, but "record" is an oblique way to convey it.

Suggest renaming to something that names the *what* rather than the side effect, e.g.:
- `module_forward` / `module_backward`
- `chunk_forward` / `chunk_backward` (matches the `ChunkRecord` / `chunk_record` argument)
- `pipeline_forward` / `pipeline_backward`

`chunk_forward` / `chunk_backward` reads most consistently with the surrounding `ChunkRecord`/`chunk_record` naming. Whatever we pick, the recording can stay a documented side effect in the docstring.

**Comment by haok1402** [comment]
@claude review

File: pithtrain/models/interface.py [hunk 6]
```diff
@@ -1,75 +1,96 @@
...
+    moe_local_idxs: Optional[torch.Tensor] = None
     expand_idx: Optional[torch.Tensor] = None
-    dedup_input_splits: Optional[List[int]] = None
-    dedup_output_splits: Optional[List[int]] = None
+    dispatch_splits: Optional[AllToAllSplits] = None
+    combine_splits: Optional[AllToAllSplits] = None
 
 
-class DecoderLayerMlpProtocol(Protocol):
+class MlpProtocol(Protocol):
```
Comment (MasterJH5574) [inline pithtrain/models/interface.py]: Let's capitalize

```suggestion
class MLPProtocol(Protocol):
```

Comment (haok1402) [reply pithtrain/models/interface.py]: Fixed.

File: pithtrain/models/gpt_oss.py (hunk 4 above)
Comment (haok1402) [reply pithtrain/models/gpt_oss.py:374]: @MasterJH5574 Good catch! Fixed.

File: pithtrain/dualpipe/execution.py (hunk 5 above)
Comment (haok1402) [reply pithtrain/dualpipe/execution.py]: @MasterJH5574 I think `record_` names the defining behavior, not a side effect: it stashes every stage's activations into `ChunkRecord` for the manual pipeline backward, which is exactly what separates it from `reference_forward` (same compute, plain autograd).

**Comment by haok1402** [comment]
> ### Consistency Review

Fixed!

File: pithtrain/modules/training.py [hunk 7]
```diff
@@ -530,28 +459,20 @@ def setup_model(
...
     torch.manual_seed(cfg.training.seed)
     torch.cuda.manual_seed_all(cfg.training.seed)
-    setup_model(cfg.training, ctx.training, cfg.distributed, ctx.distributed)
-    ctx.training.optimizers = cfg.training.optimizer(cfg.training, ctx.training)
-    ctx.training.schedulers = cfg.training.scheduler(cfg.training, ctx.training)
-    try:
-        gc.disable()
-        yield ctx.training
-    finally:
-        gc.enable()
```
Comment (MasterJH5574) [inline pithtrain/modules/training.py:557]: Maybe need to double check how this was introduced and if it's safe to remove this

File: pithtrain/operators/deepgemm_quantize.py [hunk 8]
```diff
@@ -849,16 +845,12 @@ def _fused_blockwise_transpose_fp8_batched_kernel(
...
-
-    # Store transposed FP8 data: out_t(G, K, N)
-    t_row_offs = col_start + tl.arange(0, BLOCK)
-    t_col_offs = row_start + tl.arange(0, BLOCK)
+    # Store to out_t(G, K, N) with transposed addressing rather than tl.trans, which
+    # miscompiles this fp8 tile: element (i, j) -> out_t[col_start + j, row_start + i].
     out_t_base = out_t_ptr + pid_g * stride_out_t_g
-    out_t_ptrs = out_t_base + t_row_offs[:, None] * stride_out_t_row + t_col_offs[None, :]
-    t_mask = (t_row_offs[:, None] < K) & (t_col_offs[None, :] < N)
-    tl.store(out_t_ptrs, fp8_tile_t, mask=t_mask)
```
Comment (MasterJH5574) [inline pithtrain/operators/deepgemm_quantize.py:861]: need to make sure the related unit tests can pass

File: pithtrain/modules/training.py (hunk 7 above)
Comment (haok1402) [reply pithtrain/modules/training.py:557]: @MasterJH5574 This was moved to the `pretrain_lm` file instead of deleted.

https://github.com/haok1402/pith-train/blob/dc93ac511ad26609e00a2afa7b133825b01dee09/pithtrain/tasks/pretrain_lm.py#L507

Its introduction was primarily to avoid any CPU pause that could stall DualPipe.

File: pithtrain/operators/deepgemm_quantize.py (hunk 8 above)
Comment (haok1402) [reply pithtrain/operators/deepgemm_quantize.py:861]: Reverted. We'll address Hopper FP8 in a follow-up PR.

File: pithtrain/dualpipe/execution.py (hunk 5 above)
Comment (haok1402) [reply pithtrain/dualpipe/execution.py]: Fixed. Opus recommended using `model_forward/backward` to be aligned to `layer_forward/backward`.

**Review by MasterJH5574 — APPROVED** [review]
LGTM!
